### ISSUE TITLE
Wrapping the slate plugin

### ISSUE THREAD

**Issue by Fivedark** [react-page/react-page#721]
**Is your feature request related to a problem? Please describe.**
I have created some plugins based on the slate plugin but wrapped in a component. An example for such a component could be an existing headline compontent with a property for different themes like "Title", "Sub", "Minor", ... you name it.
Slate will get the global styling from this parent component, but also allow to use the hover buttons to add a link or underline a certain part. In the toolbar you will find the controls for the headline component properties like "theme".

My solution is working but:
- I have to use some parts of the slate plugin, which are not officialy exported. Like: '@react-page/plugins-slate/lib/serialization/createInitialState' or '@react-page/plugins-slate/lib/types/component'
- I had to replace/extend the (un-)serialization to handle also the states for the parent component properties
-  I don't get TypeScript to accept my plugin type, maybe because of missing SerializationFunctions in my component?

Is there a nicer way to get this to work? 
Should the create-plugin-materialui provide factories to build own plugins based on slate?

**Additional context**
Here my example code. I removed/skipped some parts (properties, type definitions, defaults, ...) to keep it short:
Plugin:
```
import React from 'react';
import { lazyLoad } from '@react-page/core';
import SlatePlugin from '@react-page/plugins-slate';
import createInitialState from '@react-page/plugins-slate/lib/serialization/createInitialState';
import { SlateProps } from '@react-page/plugins-slate/lib/types/component';
import Component from '@react-page/plugins-slate/lib/Component';

import { serialize, unserialize } from '../../utils/slate';
import { hoverPlugins } from '../../utils/HoverToolbar';

import { defaultHeadingState } from './default/state';
import HeadingRenderer from './Renderer';

const TitleIcon = lazyLoad(() => import('@material-ui/icons/Title'));
const HeadingControls = lazyLoad(() => import('./Controls'));

const slate = SlatePlugin();

const Heading = {
  ...slate,
  Component: (props: SlateProps) => (
    <Component
      Renderer={HeadingRenderer}
      Controls={HeadingControls}
      plugins={hoverPlugins}
      serializeFunctions={{}}
      {...props}
    />
  ),
  name: 'teamshirts/content/heading',
  version: '0.0.1',
  IconComponent: <TitleIcon />,
  text: 'Heading',
  description: 'Show a Headline',
  createInitialState: () => ({
    ...createInitialState(),
    ...defaultHeadingState,
  }),
  serialize: serialize(slate.serialize),
  unserialize: unserialize(slate.unserialize),
};

export default Heading;
```
Renderer:
```
import React, { FunctionComponent } from 'react';
import Renderer from '@react-page/plugins-slate/lib/Renderer';
import Heading from '@teamshirts/components.heading';

import { HeadingControlsProps } from '../types/controls';

const HeadingRenderer: FunctionComponent<HeadingControlsProps> = props => {
  const { state, plugins } = props;

  const { theme } = state;

  return (
    <Heading
      theme={theme}
    >
      <Renderer state={state} plugins={plugins} />
    </Heading>
  );
};

export default HeadingRenderer;
```

Controls:
```
import React, { FunctionComponent } from 'react';
import { Editor } from 'slate-react';
import { BottomToolbar } from '@react-page/ui';
import FormControl from '@material-ui/core/FormControl/FormControl';
import InputLabel from '@material-ui/core/InputLabel/InputLabel';
import Select from '@material-ui/core/Select/Select';
import Input from '@material-ui/core/Input/Input';
import MenuItem from '@material-ui/core/MenuItem/MenuItem';
import FormControlLabel from '@material-ui/core/FormControlLabel/FormControlLabel';
import Switch from '@material-ui/core/Switch/Switch';
import { withStyles } from '@material-ui/styles';
import Heading from '@teamshirts/components.heading';

import { darkTheme } from '../../../../../../config/theme';
import { bottomToolbarStyles, handleChange } from '../../../utils/component';
import LeaderInput from '../../../controls/LeaderInput';
import TrailerInput from '../../../controls/TrailerInput';
import { HeadingControlsProps } from '../types/controls';

const themes = [
  'title',
  'sub',
  'minor',
];

const HeadingControls: FunctionComponent<HeadingControlsProps> = props => {
  const { state, onChange, focused, readOnly, plugins, classes } = props;

  const {
    editorState,
    theme,
  } = state;

  const onStateChange = ({ value }) =>
    onChange({ ...state, editorState: value });

  return (
    <div>
      <Heading
        theme={theme}
      >
        <Editor
          onChange={onStateChange}
          readOnly={Boolean(readOnly)}
          className="ory-plugins-content-slate-container"
          value={editorState}
          plugins={plugins}
          placeholder="Your Headline here ..."
        />
      </Heading>
      {readOnly ? null : (
        <BottomToolbar open={focused} theme={darkTheme}>
          <FormControl className={classes.formControl}>
            <InputLabel htmlFor="theme-helper">Theme</InputLabel>
            <Select
              name="theme"
              value={theme || 'title'}
              onChange={event => handleChange(event, onChange)}
              input={<Input name="theme" id="theme-helper" />}
            >
              {themes.map(themeItem => (
                <MenuItem value={themeItem} key={themeItem}>
                  {themeItem}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </BottomToolbar>
      )}
    </div>
  );
};

export default withStyles(bottomToolbarStyles)(HeadingControls);
```

**Comment by macrozone** [comment]
I don't 100% understand the question. The goal is to switch a "theme" on the whole slate area or just on a component?

> An example for such a component could be an existing headline compontent with a property for different themes like "Title", "Sub", "Minor", ... you name it.

that sound more like just using three heading types (level 1, 2 and 3) or so (?)

If you just want to have one Toolbar button, but want to display a dialog where you can customize the component (e.g. with a select-field that choses the "theme"), you can do it like in the example: https://github.com/react-page/react-page/blob/feat/uniforms-in-slate/examples/src/customSlatePlugin.tsx 

So in your case, it could look very similar:

```
import { pluginFactories } from '@react-page/plugins-slate';
import React from 'react';

export default pluginFactories.createComponentPlugin<{
  color: string;
}>({
  addHoverButton: false,
  addToolbarButton: true,
  type: 'CUSTOMHEADING',
  object: 'block',
  replaceOnRemove: 'PARAGRAPH/PARAGRAPH', // if you set this property, the plugin will replace the paragraph with a the Component below, If you don't set it, the Component will wrap whathever currently is there (typically a paragraph)
  icon: ...,
  Component: props => {
    const theme = props.data.get('theme') 
    return (
     // you can do whathever you want with that theme property, e.g. have some object where all diferent styles are defined
      <h1 style={STYLES[theme]}>{props.children}</h1>
    );
  },
  schema: {
    type: 'object',
    required: ['theme'],
    properties: {
      theme: {
        type: 'string',
     
        enum: ['fancy', 'classic', 'glorious', 'punky', 'avantgarde'],
      },
    },
  },
});
```

this would be a customizable heading. What you could also do is wrapping the current block with a `<div>` that sets a certain class. This looks very similar, but notice that it does not set `replaceOnRemove`:

```
import { pluginFactories } from '@react-page/plugins-slate';
import React from 'react';

export default pluginFactories.createComponentPlugin<{
  color: string;
}>({
  addHoverButton: false,
  addToolbarButton: true,
  type: 'THEMEBLOCK',
  object: 'block',
  icon: ...,
  Component: props => {
    const theme = props.data.get('theme') 
    return (
      <div className={`theme-${theme}`}>{props.children}</div>
    );
  },
  schema: {
    type: 'object',
    required: ['theme'],
    properties: {
      theme: {
        type: 'string',
     
        enum: ['fancy', 'classic', 'glorious', 'punky', 'avantgarde'],
      },
    },
  },
});
```

**Some more answers** 

> Should the create-plugin-materialui provide factories to build own plugins based on slate?

--> no, but slate plugin factories now use that too, if you provide a `schema` property, so you don't need to import that package

> I have to use some parts of the slate plugin, which are not officialy exported. Like: '@react-page/plugins-slate/lib/serialization/createInitialState' or '@react-page/plugins-slate/lib/types/component'

Yea, you should not do that, because that might break in future versions. Also /lib is es5 code and tree-shaking might not work properly anymore. if you really need something from there, you can send in a PR that adds that to the export

Also in your example, you seem to replace the whole slate plugin with a `Heading` plugin, which is not logical to me. Maybe there is a confusion about the plugins:

- `react-page` has different plugins, that represent a block in the editor, e.g. images, videos and text-areas
- `slate` itself is a text-editor and the `slate()`-plugin is a react-page plugin that shows a rich-text area in a block
- `slate` itself has also plugins, "slate-plugins", which can themself be blocks, inline elements or marks (e.g. `em`, `strong`).
- Some stuff make sense to be added as slate-plugins, some are more suitable as react-page plugins. Headings for example make most sense as slate-plugins, like it is in the default configuration

**Comment by macrozone** [comment]
If you want to have different slate plugins, e.g. one with only heading and a paragraph, others with links, all headings, lists, etc. then you can just use the slate plugin with different configurations:

```

const slate1 = slate([slatePlugins.headings.h1(), slatePlugins.paragraph()];
const slate2 = slate(); // all default plugins
const fancyHeading = slatePlugins.headings.h1(def => ({...def, Component: FancyHeading}))
const slate3 = slate([fancyHeading, slatePlugins.paragraph()]
```

make the slate plugin itself configurable for the user is not yet so easy, but we could also change the api a bit so that it's easier to customize e.g. with a schema-based form

**Comment by Fivedark** [comment]
Sorry, that I haven't been clear enough. I will try again:
We have a whole set of react components to build our page, like heading, text, button, accordion ... Each of this components has a list of properties to change the styling and behavior. 
Now we want to use react-page to build a landingpage configurator, to allow our marketing team to build special pages via drag-an-drop while using our set of components.
So we need for each of our components one react-page plugin with controls for each property.
This is really working fine, but for some spezial components like text or heading, I would like to use slate to allow inline typing instead of inserting the content/text via a input control. 

So my current solution described above is, to wrap my heading component around the slate editor.
I think I can't achieve that with a slate plugins or different slate instances.

Here a screenshot where the headline and the text could be changed inline but using our components.
![React-Page wrapped Slate](https://user-images.githubusercontent.com/16650470/63938096-4b471200-ca64-11e9-9282-f659a41bdd36.PNG)

**Comment by macrozone** [comment]
you could make layoutPlugins that have one slate element inside:


```
import { v4 } from "uuid";

export default createLayoutPlugin({
  Renderer: props => <h1>{props.children}</h1>,
  name: "heading",
  version: "0.0.1",
  // IconComponent: ,
  text: "Heading",
  description: "Heading",
  schema: {
     .... you could use the schema to create your complex form below
     ... altough we might could need some more layout options to it, so that the form controls float like in your example
 },
  createInitialChildren: () => {
    // wee https://github.com/react-page/react-page/issues/714
    const slateInst = new LayoutPlugin(slate([slatePlugins.emphasize(), slatePlugins.link(), slatePlugins.paragraph()]);
    return {
      id: v4(),
      rows: [
        {
          id: v4(),
          cells: [
            {
              size: 12,
              content: {
                plugin: slateInst,
                state: slateInst.createInitialState(),
              },
              id: v4(),
            },
          ],
        },
      ],
    };
  },
});

**Comment by macrozone** [comment]
or you could send in a pr, what would allow this api:

```
const mySlate = slate(defaultConfig => ({
  ...defaultConfig,
  plugins: defaultConfig.plugins, 
  Component: props => <MyWrapper><Component {...props} />
})
```

that would be more similar to the new plugin factories

**Comment by macrozone** [comment]
0.10.0-beta.3 should have this new api, see the release log

won't solve it completly, but is a step forward

**Comment by Fivedark** [comment]
Wow, thats really cool of you. Thx!

I'm really sorry, but I don't get it. I didn't get this Renderer and my own Controls to work. 
Maybe the slate plugin needs a small change? Like this:
```
  return {
    Component: (props: SlateProps) => (
      <Component
        Renderer={*settings.*Renderer}
        Controls={*settings.*Controls}
        plugins={plugins}
        translations={settings.translations}
        serializeFunctions={serializeFunctions}
        {...props}
      />
    ),
```

And do you have any idea, how to solve the serialization problem? 
Actually I have replace serialize and unserialize directly (see my first post) to also serialize also my other props and not only the slate editorState. But this behavior didn't replace the `serializeFunctions`.
```
import { SlateState } from '@react-page/plugins-slate/lib/types/state';
import { ValueJSON } from 'slate';

export const unserialize = originalUnserialize => (
  slateState: SlateState,
): SlateState => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { importFromHtml, serialized, editorState, ...rest } = slateState;

  return {
    ...originalUnserialize(slateState),
    ...rest,
  };
};

export const serialize = originalSerialize => (
  slateState: SlateState,
): { serialized: ValueJSON } => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { editorState, ...rest } = slateState;

  return {
    ...originalSerialize(slateState),
    ...rest,
  };
};
```

**Comment by macrozone** [comment]
@Fivedark can you show a more complete example? How do you use your `unserialize` and `serialize` functions? And what change do you propose in the slate plugin?

**Comment by Fivedark** [comment]
I wrapped my changes in *, but this wasn't flashy enought: I add the `setting.` to Renderer and Controls to use them from the settings and not the imported ones.

Like shown in my first post here, I added the following lines to my plugin config:
```
  serialize: serialize(slate.serialize),
  unserialize: unserialize(slate.unserialize),
```

**Comment by macrozone** [comment]
@Fivedark I need a more complete example using the `slate()` plugin. Something like:


```
slate(originalConfig => ({
   ...originalConfig,
   serialize: // what do you suggest here? 
   unserialize: // same same,
   Renderer: // what do you need here?

})
```

**Comment by Fivedark** [comment]
@macrozone I'm not sure, how it should look like, I just tried to use the API you added. But I didn't get it to work. So I had a look at your code changes and thought that there is missing something. The Renderer and so <MyWrapper> in your Example is never called, in my opinion.

To make it more clear, here your example, but MyWrapper will not render:
```
const mySlate = slate(defaultConfig => {
const DefaultRenderer = defaultConfig.Renderer;
return {
  ...defaultConfig,
  plugins: defaultConfig.plugins, 
  Renderer: props => <MyWrapper><DefaultRenderer {...props} /></MyWrapper>
}
}
```

For serialization, I think, it is maybe the best to extend the existing functions of the slate plugin to also handle other properties than the editorState.

**Comment by macrozone** [comment]
@Fivedark which version do you use?

**Comment by macrozone** [comment]
@Fivedark oooooh, sorry, sorry, sorry. There is indeed a bug! wait a moment...

**Comment by macrozone** [comment]
@Fivedark can you try 0.10.2 or higher? (0.11.0 is latest)

**Comment by Fivedark** [comment]
@macrozone Thank you! I've tested 0.11.0 and the Renderer is working now, except the properties of my wrapper component. They are lost, because of the (un)serialization methods. Is it possible to extend the methods in the slate plugin like this:

```
const unserialize = ({
  importFromHtml,
  serialized,
  editorState,
  ...rest
}: // tslint:disable-next-line:no-any
SlateState): SlateState => {
  if (serialized) {
    // tslint:disable-next-line:no-any
    return { editorState: (Value.fromJSON as any)(serialized), ...rest };
  } else if (importFromHtml) {
    return { editorState: htmlToSlate(importFromHtml), ...rest };
  } else if (editorState) {
    return { editorState, ...rest };
  }

  return { ...createInitialState(), ...rest };
};

// tslint:disable-next-line:no-any
const serialize = ({
  editorState,
  ...rest
}: SlateState): { serialized: ValueJSON } => ({
  // tslint:disable-next-line:no-any
  serialized: (editorState.toJSON as any)(editorState),
  ...rest,
});
```

**Comment by macrozone** [comment]
@Fivedark can you send in a PR?

**Comment by Fivedark** [comment]
Sorry, I'm not able to send in a PR at the moment, because I'm not able to install the react-page project on my windows machine anymore. 
The problem is the new `find` in the `clean` command in package.json. The windows `find` command use another parameter structure than the linux version, so an error is occured.

```
yarn install v1.17.3
[1/5] Resolving packages...
success Already up-to-date.
$ npm-run-all clean bootstrap build:lib
yarn run v1.17.3
$ lerna run clean && find . -name tsconfig-es.tsbuildinfo  -delete && find . -name tsconfig.tsbuildinfo  -delete
lerna notice cli v3.6.0
lerna info Executing command in 13 packages: "yarn run clean"
lerna info run Ran npm script 'clean' in '@react-page/core' in 0.4s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-divider' in 0.4s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-default-native' in 0.5s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/ui' in 0.5s:
$ rimraf "lib/**/*" && rimraf "lib-es/*"
lerna info run Ran npm script 'clean' in '@react-page/plugins-html5-video' in 0.5s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-image' in 0.6s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-spacer' in 0.6s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-video' in 0.6s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/create-plugin-materialui' in 0.6s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-background' in 0.6s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-parallax-background' in 0.6s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/plugins-slate' in 0.4s:
$ rimraf "lib" && rimraf "lib-es"
lerna info run Ran npm script 'clean' in '@react-page/renderer' in 0.4s:
$ rimraf "lib" && rimraf "lib-es"
lerna success run Ran npm script 'clean' in 13 packages in 2.4s:
lerna success - @react-page/core
lerna success - @react-page/plugins-divider
lerna success - @react-page/plugins-html5-video
lerna success - @react-page/plugins-image
lerna success - @react-page/plugins-default-native
lerna success - @react-page/plugins-slate
lerna success - @react-page/plugins-spacer
lerna success - @react-page/plugins-video
lerna success - @react-page/create-plugin-materialui
lerna success - @react-page/plugins-background
lerna success - @react-page/plugins-parallax-background
lerna success - @react-page/renderer
lerna success - @react-page/ui
FIND: Parameterformat falsch
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "clean" exited with 2.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

Even on a MINGW64 (linux based console on windows) console crashes the install process, with an unspecified error and some permission problems.

**Comment by macrozone** [comment]
@Fivedark you can comment out the `find` calls in package.json, maybe there is even a polyfill for find that works in both windows and linux/unix?

Hab leider lange nicht mehr auf windows entwickelt :-/

**Comment by Fivedark** [comment]
Sorry for my late response, but I'm working only two days a week on this project. 

To create the PR I tried to comment out the `find` calls, but got some other problems. `yarn install` crashed because it couldn't find `@react-page/core` on ui and other packages. Maybe a poblem with symlinks in windows, but how could it work a few weeks ago?

Never mind I setup a docker container to run the project inside. It is not really comfortable, because `watch` isn't working, but I was able to check if the example is still working.
